src/smartstream/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+use std::fmt;
+pub use fluvio_dataplane_protocol as dataplane;
+use dataplane::record::DefaultRecord;
+
+#[cfg(feature = "derive")]
+pub use fluvio_smartstream_derive::smartstream;
+
+pub mod memory {
+    /// Allocate memory into the module's linear memory
+    /// and return the offset to the start of the block.
+    #[no_mangle]
+    pub fn alloc(len: usize) -> *mut u8 {
+        // create a new mutable buffer with capacity `len`
+        let mut buf = Vec::with_capacity(len);
+        // take a mutable pointer to the buffer
+        let ptr = buf.as_mut_ptr();
+        // take ownership of the memory block and
+        // ensure the its destructor is not
+        // called when the object goes out of scope
+        // at the end of the function
+        std::mem::forget(buf);
+        // return the pointer so the runtime
+        // can write data at this offset
+        ptr
+    }
+
+    #[no_mangle]
+    #[allow(clippy::missing_safety_doc)]
+    pub unsafe fn dealloc(ptr: *mut u8, size: usize) {
+        let data = Vec::from_raw_parts(ptr, size, size);
+
+        std::mem::drop(data);
+    }
+}
+
+/// A trait for any type that may construct itself from a DefaultRecord.
+pub trait FromRecord: Sized {
```

Comment:
Why can't use TryFrom?

Reply:
I was taking inspiration from how actix-web's extractors work https://github.com/actix/actix-web/blob/master/src/extract.rs#L17

I think it would be really cool to eventually support something like this:

```rust
#[derive(Serialize, Deserialize)]
struct MyData {
    name: String,
    number: u32,
}

#[smartstream(filter)]
fn find_bob(data: Json<MyData>) -> bool {
    data.inner.name == "bob"
}
```

But I imagine that in order to do that we will need more control over the conversion traits

Reply:
Yes,  we should support for may conversion but it is out of scope for this issue

Reply:
Right. I'm just saying that is the rationale for why I used a custom trait rather than TryFrom/From